doc/man/dcut.1
.ds command dcut
.ds COMMAND DCUT
.\" ==========
.TH "\*[COMMAND]" 1 "2017-01-09" "Debian"
.
.\" ==========
.SH NAME
.B \*[command]
\- Debian command upload tool for the archive upload queue
.
.\" ==========
.SH SYNOPSIS
.
.SY \*[command]
.OP \-Pds
.OP \-c CONFIGFILE
.OP \-k KEYID
.OP \-m MAINTAINER
.OP \-O COMMANDFILE
\f[R][\f[]
.RB [ \-\-host ]
.IR HOSTNAME ]
.I COMMAND
.RB [ ,
.IR COMMAND ]
\f[R].\|.\|.\f[]
.YS
.
.SY \*[command]
.OP \-Pds
.OP \-c CONFIGFILE
.OP \-k KEYID
.OP \-m MAINTAINER
.OP \-O COMMANDFILE
\f[R][\f[]
.RB [ \-\-host ]
.IR HOSTNAME ]
.B \-i
.I CHANGESFILE
.YS
.
.SY \*[command]
.OP \-Pds
.OP \-c CONFIGFILE
\f[R][\f[]
.RB [ \-\-host ]
.IR HOSTNAME ]
.B \-U
.I COMMANDFILE
.YS
.
.SY \*[command]
.B \-h
.YS
.
.SY \*[command]
.B \-v
.YS
.
.\" ==========
.SH DESCRIPTION
.
.P
.B \*[command]
creates and / or uploads Debian queue commands files for a Debian
repository upload queue.
.
It is a companion for
.BR dput (1)
and uses dput configuration files (see \f[B]FILES\f[], below).
.
.P
The normal workflow is:
.
.IP \[bu]
Determine the queue control commands.
.
.IP \[bu]
Compose the batch of commands to upload.
.
.IP \[bu]
Determine the hostname of the queue host.
.
.IP \[bu]
Upload the commands to the queue host.
.
.SS Determine queue control commands
.
.P
In the default command-line syntax, parse each command-line argument
\f[I]COMMAND\f[], in the format needed by a Debian archive queue
manager.
.
Multiple queue commands must be separated by a comma
.RB ( , )
as a separate word between each command.
.
.P
See \f[B]COMMANDS\f[] below for valid queue commands.
.
.P
When the \f[I]CHANGESFILE\f[] parameter is specified, instead parse
that file and compose a \f[B]rm\f[] command for each file specified in
\f[I]CHANGESFILE\f[].
.
.P
If the
.B \-\-upload
.RB ( \-U )
option is specified, instead read the queue control commands from the
file \f[I]COMMANDFILE\f[].
.
.SS Compose the batch of commands
.
.P
The \f[B]Uploader\f[] field in the batch is set to the package
maintainer specification \f[I]MAINTAINER\f[].
.
If not specified, the value of \f[I]MAINTAINER\f[] is derived from
environment variables (see \f[B]ENVIRONMENT\f[], below).
.
.P
The \f[B]Commands\f[] field in the batch is set to the sequence of
queue control commands determined.
.
.P
If the
.B \-\-output
.RB ( \-O )
option is specified, write the batch to the file named by its
\f[I]COMMANDFILE\f[] argument.
.
Otherwise, write the batch to a temporary file and remove it after
upload.
.
.P
After writing the batch of commands to a file, sign the file with
.BR debsign (1).
.
The \f[I]MAINTAINER\f[] and \f[I]KEYID\f[] parameters, if specified,
are passed to the
.B debsign
command.
.
.P
If the
.B \-\-upload
.RB ( \-U )
option is specified, do not compose the batch of commands; instead use
the content of its \f[I]COMMANDFILE\f[] argument as the name of the
command file to upload to the queue host.
.
.SS Determine the queue host
.
The \f[I]HOSTNAME\f[] must match one of the named hosts in the
configuration (see
.BR dput.cf (5)).
.
The value is determined as follows:
.
.IP \[bu]
If the \f[B]\-\-host\f[] option is specified, use that option's
\f[I]HOSTNAME\f[] argument.
.
.IP \[bu]
If the first non-option argument is not a valid queue command, that is
interpreted as the \f[I]HOSTNAME\f[].
.
.IP \[bu]
Otherwise, \f[I]HOSTNAME\f[] defaults to the value of the
.B default_host_main
configuration parameter.
.
.SS Upload queue commands
.
.P
Upload the composed batch of queue commands to the queue host
\f[I]HOSTNAME\f[], using the parameters in the
.BR dput.cf (5)
configuration for that host.
.
.P
If the
.B \-\-output
.RB ( \-O )
option is specified, do not upload the commands to the queue host.
Instead, leave the composed file \f[I]COMMANDFILE\f[] on the
filesystem.
.
.\" ==========
.SH COMMANDS
.
.P
Presently supported commands are:
.
.TP
.B rm
Remove files from the UploadQueue directories.
.
\f[B]\*[command]\f[] generates commands that search the subdirectories
automatically, but you can also specify \f[B]\-\-nosearchdirs\f[] and
then target individual UploadQueue directories, i.e. either filenames
without path for regular incoming or
\f[B]DELAYED/\f[]\f[I]DAYS\f[]\f[B]-day\f[]\f[B]/\f[]\f[I]filename\f[].
.
Wildcards are accepted.
.
.TP
.BI reschedule " CHANGESFILE DAYS" -day
Move an upload from any \f[B]DEFERRED\f[] queue to
.IB DAYS -day
(use \fB0-day\fR for immediate processing).
.
.TP
.BI cancel " CHANGESFILE"
Cancels an upload from \fBDEFERRED\fR.
.
.P
Both \fBreschedule\fR and \fBcancel\fR take a complete basename
of a changes file as argument and operate on the whole upload,
they process the changes and all files mentioned by it.
.
.P
The authoritative documentation for the commands interface
can be found at
.UR ftp://ftp.upload.debian.org/pub/UploadQueue/README
.UE .
.
.\" ==========
.SH OPTIONS
.
.TP
.BI \-c " CONFIGFILE"
.TQ
.BI \-\-config " CONFIGFILE"
Use the file \f[I]CONFIGFILE\f[] for configuration.
.
See
.BR dput.cf (5)
for the specification of this file format.
.
.TP
.B \-d
.TQ
.B \-\-debug
Display debugging messages.
.
.TP
.B \-h
.TQ
.B \-\-help
Display help information, then exit.
.
.TP
.BI \-\-host " HOSTNAME"
Specify the \f[I]HOSTNAME\f[] unambiguously.
.
This allows naming a host that may be ambiguous with a queue command.
.
.TP
.BI \-m " MAINTAINER"
.TQ
.BI \-\-maintainer " MAINTAINER"
Use \f[I]MAINTAINER\f[] for the \f[B]Uploader\f[] field and GnuPG key
selection.
.
This has no effect when \f[B]\-\-upload\f[] is used.
.
.TP
.BI \-k " KEYID"
.TQ
.BI \-\-keyid " KEYID"
Specify the GnuPG key id \f[I]KEYID\f[] to pass to the
.BR debsign (1)
command.
.
This has no effect when \f[B]\-\-upload\f[] is used.
.
.TP
.BI \-O " COMMANDFILE"
.TQ
.BI \-\-output " COMMANDFILE"
Write commands file to \f[I]COMMANDFILE\f[] instead of uploading.
.
You should not specify a \f[I]HOSTNAME\f[] when using this option.
.
Also, this option should not be used with \f[B]\-\-upload\f[].
.
.TP
.B \-P
.TQ
.B \-\-passive
Use passive mode for FTP.
.
If not specified, default for FTP is active mode.
.
.TP
.B \-s
.TQ
.B \-\-simulate
Do not actually upload to the queue host.
.
.TP
.BI \-U " COMMANDFILE"
.TQ
.BI \-\-upload " COMMANDFILE"
Instead of composing a new queue command file, upload the existing
file \f[I]COMMANDFILE\f[].
.
Note: No checks (syntax or presence/validity of signature) are
currently performed.
.
.TP
.BI \-i " CHANGESFILE"
.TQ
.BI \-\-input " CHANGESFILE"
Do not parse queue commands from the command line. Instead, read the
package changes file \f[I]CHANGESFILE\f[] and compose a \f[B]rm\f[]
command for each file specified in \f[I]CHANGESFILE\f[].
.
This has no effect when \f[B]\-\-upload\f[] is used.
.
.TP
.B \-v
.TQ
.B \-\-version
Display version information, then exit.
.
.\" ==========
.SH ENVIRONMENT
.
.TP
.B DEBEMAIL
.TQ
.B EMAIL
.
If \f[I]MAINTAINER\f[] is not specified, use the first of these which
has a value to determine the email address portion of
\f[I]MAINTAINER\f[].
.
.TP
.B DEBFULLNAME
.
If \f[I]MAINTAINER\f[] is not specified, use this value to determine
the full name portion of \f[I]MAINTAINER\f[].
.
.TP
.B USER
.
If no username is specified for authenticating to the remote host, use
the value of this variable.
.
.\" ==========
.SH FILES
.
.TP
Configuration
.
See
.BR dput.cf (5)
for the files to configure
.B \*[command].
.
.\" ==========
.SH EXAMPLES
.
.P
To upload a GPG signed commands file that removes dput_0.9.2.33_all.deb
file from the queue, one would have to do:
.
.P
.EX
$ \*[command] rm dput_0.9.2.33_all.deb
.EE
.
.P
This creates and uploads a commands file that looks like the following:
.
.P
.EX
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA1

Uploader: Foo Bar <foo@bar.org>
Commands:
 rm dput_0.9.2.33_all.deb


-----BEGIN PGP SIGNATURE-----
Version: GnuPG v1.4.9 (GNU/Linux)

iEYEARECAAYKAkisYx0ACgkQ4eu+pS04mIfwLQCg85XVpae/LHLrbvzywcQNt9PS
ztsAoMgw6HjcI9flfmwjgMiapcu379Pu
=fjp0
-----END PGP SIGNATURE-----
.EE
.
.P
To upload a GPG signed commands file that removes files named in
dput_0.9.4_i386.changes from the queue, one would have to do:
.
.P
.EX
$ \*[command] \-\-input dput_0.9.4_i386.changes
.EE
.
.P
This creates and uploads a commands file that looks like the following:
.
.P
.EX
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA1

Uploader: Foo Bar <foo@bar.org>
Commands:
 rm \-\-searchdirs dput_0.9.4.dsc
 rm \-\-searchdirs dput_0.9.4.tar.gz
 rm \-\-searchdirs dput_0.9.4_all.deb


-----BEGIN PGP SIGNATURE-----
Version: GnuPG v1.4.9 (GNU/Linux)

iEYEARECAAYKAkisYx0ACgkQ4eu+pS04mIfwLQCg85XVpae/LHLrbvzywcQNt9PS
ztsAoMgw6HjcI9flfmwjgMiapcu379Pu
=fjp0
-----END PGP SIGNATURE-----
.EE
.
.P
If you've uploaded packages with the \f[B]\-\-delayed\f[] option
(uploaded to \f[B]DEFERRED\f[] queue), then use the \f[B]cancel\f[]
command with a .changes file:
.
.P
.EX
$ \*[command] cancel dput_0.9.4_i386.changes
.EE
.
.P
This creates and uploads a commands file just like the other commands.
.
.\" ==========
.SH SEE ALSO
.
.BR dput (1),
.BR dput.cf (5),
.BR debsign (1)
.
.P
.UR ftp://ftp.upload.debian.org/pub/UploadQueue/README
Read Me document for the Debian upload queue
.UE .
.
.\" Copyright © 2016–2018 Ben Finney <bignose@debian.org>
.
.\" This is free software: you may copy, modify, and/or distribute this work
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; version 3 of that license or any later version.
.\" No warranty expressed or implied. See the file ‘LICENSE.GPL-3’ for details.
.
.\" Local variables:
.\" coding: utf-8
.\" mode: nroff
.\" End:
.\" vim: fileencoding=utf-8 filetype=nroff :


debian/changelog
dput (1.0.2ubuntu2) disco; urgency=medium

  * dput/methods/sftp.py:
    - the sftp method now accepts a -p port parameter, thanks Valentyna
      (lp: #1812174)

 -- Sebastien Bacher <seb128@ubuntu.com>  Fri, 18 Jan 2019 11:52:10 +0100

dput (1.0.2ubuntu1) cosmic; urgency=low

  * Merge from Debian unstable.  Remaining changes:
    - dput:
      + Recognize 0ubuntu1 as a debian version that requires
        orig.tar.gz to be included in the upload.
      + Correctly handle cases where a non-existant host or no host at all is
        supplied when -o option is passed. (Thanks to David Futcher for
        spotting)
      + Be more careful about setting a variable in a section that does not
        exist in host argument handling.
    - dput.1: Updated to document host argument feature and sftp support.
    - dput.cf:
      + Set 'default_main_host = ubuntu'
      + Set 'progress_indicator = 2'
      + Updated ppa stanza to make use of argument support.
    - dput.cf.5: Updated to note support for sftp transport and host args.
    - dput.cf: Switch ubuntu stanza to upload to "/ubuntu" rather than "/"
      (LP: #1340130).
    - dput.cf: Drop trailing "/ubuntu" from ppa stanza, to support the new
      form of the upload path needed for PPAs based on derived distributions
      (LP: #1340130).
    - Implement a new sftp method that connects via openssh and then
      uses paramiko's sftp support. Some code copied from bzr.

 -- Julian Andres Klode <juliank@ubuntu.com>  Fri, 27 Apr 2018 13:42:28 +0200

dput (1.0.2) unstable; urgency=medium

  * The “Rezaul Karim Siddique” release.
  * Parse custom port value as integer from FQDN value text.
    Closes: bug#876382. Thanks to Marc Haber for the report.
  * Declare “Standards-Version: 4.1.3”. No additional changes needed.
  * Declare Debhelper compatibility level 11.
  * Update copyright years.
  * Update reference URLs to HTTPS scheme, where appropriate.
  * Define custom report behaviour for Reportbug.
  * Declare EditorConfig settings for the code base.
  * Change ‘incoming’ for ‘mentors’ host in default configuration.
    Closes: bug#891268. Thanks to Mattia Rizzolo for the report and patch.

 -- Ben Finney <bignose@debian.org>  Sun, 25 Feb 2018 11:07:14 +1100

dput (1.0.1ubuntu1) bionic; urgency=low

  * Merge from Debian unstable.  Remaining changes:
    - dput:
      + Recognize 0ubuntu1 as a debian version that requires
        orig.tar.gz to be included in the upload.
      + Correctly handle cases where a non-existant host or no host at all is
        supplied when -o option is passed. (Thanks to David Futcher for
        spotting)
      + Be more careful about setting a variable in a section that does not
        exist in host argument handling.
    - dput.1: Updated to document host argument feature and sftp support.
    - dput.cf:
      + Set 'default_main_host = ubuntu'
      + Set 'progress_indicator = 2'
      + Updated ppa stanza to make use of argument support.
    - dput.cf.5: Updated to note support for sftp transport and host args.
    - dput.cf: Switch ubuntu stanza to upload to "/ubuntu" rather than "/"
      (LP: #1340130).
    - dput.cf: Drop trailing "/ubuntu" from ppa stanza, to support the new
      form of the upload path needed for PPAs based on derived distributions
      (LP: #1340130).
  * Implement a new sftp method that connects via openssh and then
    uses paramiko's sftp support. Some code copied from bzr.

 -- Julian Andres Klode <juliank@ubuntu.com>  Thu, 11 Jan 2018 09:51:33 +0100

dput (1.0.1) unstable; urgency=medium

  * Upload to “unstable”.
    Thanks to testers of the Python 3 migration.
  * Declare “Standards-Version: 4.1.0”. No additional changes needed.
  * Use current HTTPS URL for copyright format reference.
  * Remove declaration of Python 2 support.

 -- Ben Finney <bignose@debian.org>  Sat, 09 Sep 2017 18:51:25 +1000

dput (1.0.0) experimental; urgency=medium

  * The “Nazimuddin Samad” release.
  * Convert code base to run on Python 3.
    Closes: bug#830879. Thanks to Emilio Pozuelo Monfort for the report.
  * Remove the obsolete conffile ‘/etc/bash_completion.d/dput’.
    Closes: bug#866230. Thanks to Laurent Bigonville for the report.
  * Defer getting the uploader until after processing command-line options.
  * Declare “Standards-Version: 4.0.1”. No additional changes required.
  * Revise documentation to remove outdated material.
  * Upgrade license grant to GNU GPL v3-or-later for the whole work.
  * Remove unused GPL v2 license text.

 -- Ben Finney <bignose@debian.org>  Mon, 07 Aug 2017 10:10:47 +1000

dput (0.12.1) unstable; urgency=medium

  * The “Imran Sarker” release.
  * Specify current recommended archive host for security updates.
    Closes: bug#863348. Thanks to Ansgar Burchardt for the report.
  * Specify current VCS for this code base.


dput/dcut.py
#! /usr/bin/python3
#
# dput/dcut.py
# Part of ‘dput’, a Debian package upload toolkit.
#
# This is free software, and you are welcome to redistribute it under
# certain conditions; see the end of this file for copyright
# information, grant of license, and disclaimer of warranty.

""" dcut — Debian command upload tool. """

import os
import pwd
import shutil
import string
import subprocess
import sys
import tempfile
import textwrap
import time

from . import dput
from .helper import dputhelper


validcommands = ('rm', 'cancel', 'reschedule')


def make_usage_message():
    """ Make the program usage help message. """
    text = textwrap.dedent("""\
        Usage: %s [options] [host] command [, command]
         Supported options (see man page for long forms):
           -c file Config file to parse.
           -d      Enable debug messages.
           -h      Display this help message.
           -s      Simulate the commands file creation only.
           -v      Display version information.
           -m maintaineraddress
                   Use maintainer information in "Uploader:" field.
           -k keyid
                   Use this keyid for signing.
           -O file Write commands to file.
           -U file Upload specified commands file (presently no checks).
           -i changes
                   Upload a commands file to remove files listed in .changes.
         Supported commands: mv, rm
           (No paths or command-line options allowed on ftp-master.)
        """) % (dputhelper.get_progname(sys.argv))
    return text


def get_uploader_from_system(options):
    """ Get the system value for the ‘uploader’ option. """
    uploader = None

    # check environment for maintainer
    if options['debug']:
        sys.stdout.write(
                "D: trying to get maintainer email from environment\n")

    if 'DEBEMAIL' in os.environ:
        if os.environ['DEBEMAIL'].find('<') < 0:
            uploader = os.environ.get("DEBFULLNAME", '')
            if uploader:
                uploader += ' '
            uploader += '<%s>' % (os.environ['DEBEMAIL'])
        else:
            uploader = os.environ['DEBEMAIL']
        if options['debug']:
            sys.stdout.write(
                    "D: Uploader from env: %s\n" % (uploader))
    elif 'EMAIL' in os.environ:
        if os.environ['EMAIL'].find('<') < 0:
            uploader = os.environ.get("DEBFULLNAME", '')
            if uploader:
                uploader += ' '
            uploader += '<%s>' % (os.environ['EMAIL'])
        else:
            uploader = os.environ['EMAIL']
        if options['debug']:
            sys.stdout.write(
                    "D: Uploader from env: %s\n" % (uploader))
    else:
        if options['debug']:
            sys.stdout.write("D: Guessing uploader\n")
        pwrec = pwd.getpwuid(os.getuid())
        username = pwrec[0]
        fullname = pwrec[4].split(',')[0]
        try:
            hostname = open('/etc/mailname').read().strip()
        except IOError:
            hostname = ''
        if not hostname:
            if options['debug']:
                sys.stdout.write(
                        "D: Guessing uploader: /etc/mailname was a failure\n")
            hostname_subprocess = subprocess.Popen(
                    "/bin/hostname --fqdn",
                    shell=True, stdout=subprocess.PIPE)
            hostname_stdout = dputhelper.make_text_stream(
                    hostname_subprocess.stdout)
            hostname = hostname_stdout.read().strip()
        if hostname:
            uploader = (
                    "%s <%s@%s>" % (fullname, username, hostname))
            if options['debug']:
                sys.stdout.write(
                        "D: Guessed uploader: %s\n" % (uploader))
        else:
            if options['debug']:
                sys.stdout.write("D: Couldn't guess uploader\n")

    return uploader


def getoptions():
    # seed some defaults
    options = {
            'debug': 0, 'simulate': 0, 'config': None, 'host': None,
            'uploader': None, 'keyid': None, 'passive': 0,
            'filetocreate': None, 'filetoupload': None, 'changes': None}
    progname = dputhelper.get_progname(sys.argv)
    version = dputhelper.get_distribution_version()

    # enable debugging very early
    if ('-d' in sys.argv[1:] or '--debug' in sys.argv[1:]):
        options['debug'] = 1
        sys.stdout.write("D: %s %s\n" % (progname, version))

    # parse command line arguments
    (opts, arguments) = dputhelper.getopt(
            sys.argv[1:],
            'c:dDhsvm:k:PU:O:i:', [
                'config=', 'debug',
                'help', 'simulate', 'version', 'host=',
                'maintainteraddress=', 'keyid=',
                'passive', 'upload=', 'output=', 'input='
                ])

    for (option, arg) in opts:
        if options['debug']:
            sys.stdout.write(
                    'D: processing arg "%s", option "%s"\n' % (option, arg))
        if option in ('-h', '--help'):
            sys.stdout.write(make_usage_message())
            sys.exit(0)
        elif option in ('-v', '--version'):
            sys.stdout.write("%s %s\n" % (progname, version))
            sys.exit(0)
        elif option in ('-d', '--debug'):
            options['debug'] = 1
        elif option in ('-c', '--config'):
            options['config'] = arg
        elif option in ('-m', '--maintaineraddress'):
            options['uploader'] = arg
        elif option in ('-k', '--keyid'):
            options['keyid'] = arg
        elif option in ('-s', '--simulate'):
            options['simulate'] = 1
        elif option in ('-P', '--passive'):
            options['passive'] = 1
        elif option in ('-U', '--upload'):
            options['filetoupload'] = arg
        elif option in ('-O', '--output'):
            options['filetocreate'] = arg
        elif option == '--host':
            options['host'] = arg
        elif option in ('-i', '--input'):
            options['changes'] = arg
        else:
            sys.stderr.write(
                    "%s internal error: Option %s, argument %s unknown\n"
                    % (progname, option, arg))
            sys.exit(1)

    if not options['host'] and arguments and arguments[0] not in validcommands:
        options['host'] = arguments[0]
        if options['debug']:
            sys.stdout.write(
                    'D: first argument "%s" treated as host\n'
                    % (options['host']))
        del arguments[0]

    if not options['uploader']:
        options['uploader'] = get_uploader_from_system(options)

    # we don't create command files without uploader
    if (
            not options['uploader']
            and (options['filetoupload'] or options['changes'])):
        sys.stderr.write(
                "%s error: command file cannot be created"
                " without maintainer email\n"
                % progname)
        sys.stderr.write(
                '%s        please set $DEBEMAIL, $EMAIL'
                ' or use the "-m" option\n'
                % (len(progname) * ' '))
        sys.exit(1)

    return options, arguments


def parse_queuecommands(arguments, options, config):
    commands = []
    # want to consume a copy of arguments
    arguments = arguments[:]
    arguments.append(0)
    curarg = []
    while arguments:
        if arguments[0] in validcommands:
            curarg = [arguments[0]]
            if arguments[0] == 'rm':
                if len(arguments) > 1 and arguments[1] == '--nosearchdirs':
                    del arguments[1]
                else:
                    curarg.append('--searchdirs')
        else:
            if not curarg and arguments[0] != 0:
                sys.stderr.write(
                        'Error: Could not parse commands at "%s"\n'
                        % (arguments[0]))
                sys.exit(1)
            if str(arguments[0])[-1] in (',', ';', 0):
                curarg.append(arguments[0][0:-1])
                arguments[0] = ','
            if arguments[0] in (',', ';', 0) and curarg:
                # TV-TODO: syntax check for #args etc.
                if options['debug']:
                    sys.stdout.write(
                            'D: Successfully parsed command "%s"\n'
                            % (' '.join(curarg)))
                commands.append(' '.join(curarg))
                curarg = []
            else:
                # TV-TODO: maybe syntax check the arguments here
                curarg.append(arguments[0])
        del arguments[0]
    if not commands:
        sys.stderr.write("Error: no arguments given, see dcut -h\n")
        sys.exit(1)
    return commands


def write_commands(commands, options, config, tempdir):
    """ Write a file of commands for the upload queue daemon.

        :param commands: Commands to write, as a sequence of text
            strings.
        :param options: Program configuration, as a mapping of options
            `{name: value}`.
        :param config: `ConfigParser` instance for this application.
        :param tempdir: Filesystem path to directory for temporary files.
        :return: Filesystem path of file which was written.

        Write the specified sequence of commands to a file, in the
        format required for the Debian upload queue management daemon.

        Once writing is finished, the file is signed using the
        'debsign' command.

        If not specified in the configuration option 'filetocreate', a
        default filename is generated. In either case, the resulting
        filename is returned.

        """
    progname = dputhelper.get_progname(sys.argv)
    if options['filetocreate']:
        filename = options['filetocreate']
    else:
        translationorig = (
                str('').join(map(chr, range(256)))
                + string.ascii_letters + string.digits)
        translationdest = 256 * '_' + string.ascii_letters + string.digits
        translationmap = str.maketrans(translationorig, translationdest)
        uploadpartforname = options['uploader'].translate(translationmap)
        filename = (
                progname + '.%s.%d.%d.commands' %
                (uploadpartforname, int(time.time()), os.getpid()))
        if tempdir:
            filename = os.path.join(tempdir, filename)
    f = open(filename, "w")
    f.write("Uploader: %s\n" % options['uploader'])
    f.write("Commands:\n %s\n\n" % ('\n '.join(commands)))
    f.close()
    debsign_cmdline = ['debsign']
    debsign_cmdline.append('-m%s' % options['uploader'])
    if options['keyid']:
        debsign_cmdline.append('-k%s' % options['keyid'])
    debsign_cmdline.append('%s' % filename)
    if options['debug']:
        sys.stdout.write("D: calling debsign: %s\n" % debsign_cmdline)
    try:
        subprocess.check_call(debsign_cmdline)
    except subprocess.CalledProcessError:
        sys.stderr.write("Error: debsign failed.\n")
        sys.exit(1)
    return filename


def upload_stolen_from_dput_main(
        host, upload_methods, config, debug, simulate,
        files_to_upload, ftp_passive_mode):
    """ Upload files to the host.

        :param host: Configuration host name.
        :param upload_methods: Mapping of {method_name: callable}.
        :param config: `ConfigParser` instance for this application.
        :param debug: If true, enable debugging output.
        :param simulate: If true, simulate the upload only.
        :param files_to_upload: Collection of file names to upload.
        :param ftp_passive_mode: If true, enable FTP passive mode.
        :return: ``None``.

        Upload the specified files to the host, using the method and
        credentials from the configuration for the host.

        """
    # Messy, yes. But it isn't referenced by the upload method anyway.
    if config.get(host, 'method') == 'local':
        fqdn = 'localhost'
    else:
        fqdn = config.get(host, 'fqdn')

    # Check the upload methods that we have as default and per host
    if debug:
        sys.stdout.write(
                "D: Default Method: %s\n" % config.get('DEFAULT', 'method'))
    if config.get('DEFAULT', 'method') not in upload_methods:
        sys.stderr.write(
                "Unknown upload method: %s\n"
                % config.get('DEFAULT', 'method'))
        sys.exit(1)
    if debug:
        sys.stdout.write("D: Host Method: %s\n" % config.get(host, 'method'))
    if config.get(host, 'method') not in upload_methods:
        sys.stderr.write(
                "Unknown upload method: %s\n" % config.get(host, 'method'))
        sys.exit(1)

    # Inspect the Config and set appropriate upload method
    if not config.get(host, 'method'):
        method = config.get('DEFAULT', 'method')
    else:
        method = config.get(host, 'method')

    # Check now the login and redefine it if needed
    if (
            config.has_option(host, 'login')
            and config.get(host, 'login') != 'username'):
        login = config.get(host, 'login')
    elif (
            config.has_option('DEFAULT', 'login')
            and config.get('DEFAULT', 'login') != 'username'):
        login = config.get('DEFAULT', 'login')
    else:
        # Try to get the login from the enviroment
        if 'USER' in os.environ:
            login = os.environ['USER']
        else:
            sys.stdout.write("$USER not set, will use login information.\n")
            # Else use the current username
            login = pwd.getpwuid(os.getuid())[0]
            if debug:
                sys.stdout.write("D: User-ID: %s\n" % os.getuid())
        if debug:
            sys.stdout.write(
                    "D: Neither host %s nor default login used. Using %s\n"
                    % (host, login))
    if debug:
        sys.stdout.write("D: Login to use: %s\n" % login)

    incoming = config.get(host, 'incoming')
    # Do the actual upload
    if not simulate:
        if debug:
            sys.stdout.write("D: FQDN: %s\n" % fqdn)
            sys.stdout.write("D: Login: %s\n" % login)
            sys.stdout.write("D: Incoming: %s\n" % incoming)
        if method == 'ftp':
            ftp_mode = config.getboolean(host, 'passive_ftp')
            if ftp_passive_mode == 1:
                ftp_mode = 1
            if ftp_mode == 1:
                if debug:
                    if ftp_passive_mode == 1:
                        sys.stdout.write("D: Using passive ftp\n")
                    else:
                        sys.stdout.write("D: Using active ftp\n")
            upload_methods[method](
                    fqdn, login, incoming,
                    files_to_upload, debug, ftp_mode)
        elif method == 'scp':
            if debug and config.getboolean(host, 'scp_compress'):
                sys.stdout.write("D: Setting compression for scp\n")
            scp_compress = config.getboolean(host, 'scp_compress')
            ssh_config_options = [
                    y for y in (
                        x.strip() for x in
                        config.get(host, 'ssh_config_options').split('\n'))
                    if y]
            upload_methods[method](
                    fqdn, login, incoming,
                    files_to_upload, debug, scp_compress, ssh_config_options)
        else:
            upload_methods[method](
                    fqdn, login, incoming,
                    files_to_upload, debug, 0)
    # Or just simulate it.
    else:
        for file in files_to_upload:
            sys.stderr.write(
                    "Uploading with %s: %s to %s:%s\n"
                    % (method, file, fqdn, incoming))
            subprocess.call("cat %s" % file, shell=True)


def dcut():
    options, arguments = getoptions()
    if options['debug']:
        sys.stdout.write('D: calling dput.read_configs\n')
    config = dput.read_configs(options['config'], options['debug'])
    if (
            not options['host']
            and config.has_option('DEFAULT', 'default_host_main')):
        options['host'] = config.get('DEFAULT', 'default_host_main')
        if options['debug']:
            sys.stdout.write(
                    'D: Using host "%s" (default_host_main)\n'
                    % (options['host']))
        if not options['host']:
            options['host'] = 'ftp-master'
            if options['debug']:
                sys.stdout.write(
                        'D: Using host "%s" (hardcoded)\n'
                        % (options['host']))
    tempdir = None
    filename = None
    progname = dputhelper.get_progname(sys.argv)
    try:
        if not (options['filetoupload'] or options['filetocreate']):
            tempdir = tempfile.mkdtemp(prefix=progname + '.')
        if not options['filetocreate']:
            if not options['host']:
                sys.stdout.write(
                        "Error: No host specified"
                        " and no default found in config\n")
                sys.exit(1)
            if not config.has_section(options['host']):
                sys.stdout.write(
                        "No host %s found in config\n" % (options['host']))
                sys.exit(1)
            else:
                if config.has_option(options['host'], 'allow_dcut'):
                    dcut_allowed = config.getboolean(
                            options['host'], 'allow_dcut')
                else:
                    dcut_allowed = config.getboolean('DEFAULT', 'allow_dcut')
                if not dcut_allowed:
                    sys.stdout.write(
                            'Error: dcut is not supported'
                            ' for this upload queue.\n')
                    sys.exit(1)
        if options['filetoupload']:
            if arguments:
                sys.stdout.write(
                        'Error: cannot take commands'
                        ' when uploading existing file,\n'
                        '       "%s" found\n' % (' '.join(arguments)))
                sys.exit(1)
            commands = None
            filename = options['filetoupload']
            if not filename.endswith(".commands"):
                sys.stdout.write(
                        'Error: I\'m insisting on the .commands extension,'
                        ' which\n'
                        '       "%s" doesnt seem to have.\n' % filename)
            # TV-TODO: check file to be readable?
        elif options['changes']:
            parse_changes = dput.parse_changes
            removecommands = create_commands(options, config, parse_changes)
            filename = write_commands(removecommands, options, config, tempdir)
        else:
            commands = parse_queuecommands(arguments, options, config)
            filename = write_commands(commands, options, config, tempdir)
        if not options['filetocreate']:
            dput.import_upload_functions()
            upload_methods = dput.import_upload_functions()
            upload_stolen_from_dput_main(
                    options['host'], upload_methods, config,
                    options['debug'], options['simulate'],
                    [filename], options['passive'])
    finally:
        # we use sys.exit, so we need to clean up here
        if tempdir:
            shutil.rmtree(tempdir)


def create_commands(options, config, parse_changes):
    """ Get the removal commands from a package changes file.

        Parse the specified ‘foo.changes’ file and returns commands to
        remove files named in it.

        """
    changes_file = options['changes']
    if options['debug']:
        sys.stdout.write(
                "D: Parsing changes file (%s) for files to remove\n"
                % changes_file)
    try:
        chg_fd = open(changes_file, 'r')
    except IOError:
        sys.stdout.write("Can't open changes file: %s\n" % changes_file)
        sys.exit(1)
    the_changes = parse_changes(chg_fd)
    chg_fd.close
    removecommands = ['rm --searchdirs ' + os.path.basename(changes_file)]
    for file in the_changes['files'].strip().split('\n'):
        # filename only
        fn = file.split()[4]
        rm = 'rm --searchdirs ' + fn
        if options['debug']:
            sys.stdout.write("D: Will remove %s with '%s'\n" % (fn, rm))
        removecommands.append(rm)
    return removecommands


if __name__ == "__main__":
    try:
        dcut()
    except dputhelper.DputException as e:
        sys.stderr.write("%s\n" % e)
        sys.exit(1)


# Copyright © 2015–2018 Ben Finney <bignose@debian.org>
# Copyright © 2008–2013 Y Giridhar Appaji Nag <appaji@debian.org>
# Copyright © 2004–2009 Thomas Viehmann <tv@beamnet.de>
# Copyright © 2000–2004 Christian Kurz <shorty@debian.org>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.


# Local variables:
# coding: utf-8
# mode: python
# End:
# vim: fileencoding=utf-8 filetype=python :
